runtime/vm/jvminit.c

```diff
@@ -668,27 +668,41 @@ freeJavaVM(J9JavaVM * vm)
 		vm->cifArgumentTypesCache = NULL;
 	}
 
-	/* Clean up all resources created via allocateUpcallThunkMemory, including the generatered thunk
-	 * and the corresponding metadata of each entry stored in the hashtable.
+	/* Empty the thunk heap list by cleaning up all resources created via allocateUpcallThunkMemory,
+	 * including the generated thunk and the corresponding metadata of each entry in the hashtable.
 	 * See UpcallThunkMem.cpp for details.
 	 */
-	if (NULL != vm->thunkHeapWrapper) {
-		J9UpcallThunkHeapWrapper *thunkHeapWrapper = vm->thunkHeapWrapper;
-		J9PortVmemIdentifier vmemID = thunkHeapWrapper->vmemID;
-		UDATA byteAmount = j9vmem_supported_page_sizes()[0];
+	if (NULL != vm->thunkHeapHead) {
+		J9UpcallThunkHeapList *thunkHeapHead = vm->thunkHeapHead;
+		J9UpcallThunkHeapList *thunkHeapNode = NULL;
 
-		if (NULL != thunkHeapWrapper->metaDataHashTable) {
-			J9HashTable *metaDataHashTable = thunkHeapWrapper->metaDataHashTable;
+		if (NULL != thunkHeapHead->metaDataHashTable) {
+			J9HashTable *metaDataHashTable = thunkHeapHead->metaDataHashTable;
 			UDATA entryCount = hashTableGetCount(metaDataHashTable);
 			if (entryCount > 0) {
 				hashTableForEachDo(metaDataHashTable, (J9HashTableDoFn)freeUpcallMetaDataDoFn, NULL);
 			}
 			hashTableFree(metaDataHashTable);
 			metaDataHashTable = NULL;
 		}
-		j9vmem_free_memory(vmemID.address, byteAmount, &vmemID);
-		j9mem_free_memory(thunkHeapWrapper);
-		thunkHeapWrapper = NULL;
+
+		thunkHeapNode = thunkHeapHead->next;
+		thunkHeapHead->next = NULL;
+		thunkHeapHead = thunkHeapNode;
+		while (NULL != thunkHeapHead) {
+			J9UpcallThunkHeapWrapper *thunkHeapWrapper = thunkHeapHead->thunkHeapWrapper;
+			if (NULL != thunkHeapWrapper) {
+				J9PortVmemIdentifier vmemID = thunkHeapWrapper->vmemID;
+				UDATA byteAmount = j9vmem_supported_page_sizes()[0];
```

Comment:
I think this page-size query should be before (outside) the loop.

Reply:
Agreed and moved the query out before the loop.